src/NServiceBus.AcceptanceTests/Core/DependencyInjection/When_defining_same_dependencies_in_endpoints.cs

```diff
@@ -0,0 +1,136 @@
+namespace NServiceBus.AcceptanceTests.Core.DependencyInjection;
+
+using System;
+using System.Collections.Concurrent;
+using System.Linq;
+using System.Threading.Tasks;
+using AcceptanceTesting;
+using AcceptanceTesting.Support;
+using EndpointTemplates;
+using Features;
+using Microsoft.Extensions.DependencyInjection;
+using NUnit.Framework;
+
+public class When_defining_same_dependencies_in_endpoints : NServiceBusAcceptanceTest
+{
+    [Test]
+    public async Task Should_be_isolated_except_global_shared_once()
+    {
+        var result = await Scenario.Define<Context>()
+            .WithComponent(new CustomComponent())
+            .WithEndpoint<WithSameDependenciesEndpoint>(b =>
+                b.Services(static services => services.AddSingleton<IDependency, MyDependency>())
+                    .CustomConfig(c => c.OverrideLocalAddress("DeeplyNestedDependenciesEndpoint1"))
+                    .When((session, c) => session.Send("DeeplyNestedDependenciesEndpoint1", new SomeMessage1())))
+            .WithEndpoint<WithSameDependenciesEndpoint>(b =>
+                b.Services(static services => services.AddSingleton<IDependency, MyDependency>())
+                    .CustomConfig(c => c.OverrideLocalAddress("DeeplyNestedDependenciesEndpoint2"))
+                    .When((session, c) => session.Send("DeeplyNestedDependenciesEndpoint2", new SomeMessage1())))
+            .Done(c => c.Dependencies.Count == 2)
+            .Run();
+
+        using (Assert.EnterMultipleScope())
+        {
+            Assert.That(result.Dependencies, Has.Count.EqualTo(2), "Dependencies should have been resolved");
+            Assert.That(result.Dependencies.ElementAt(0), Is.Not.SameAs(result.Dependencies.ElementAt(1)));
+            Assert.That(result.Dependencies.ElementAt(0).Dependency, Is.Not.SameAs(result.Dependencies.ElementAt(1).Dependency));
+            Assert.That(result.Dependencies.ElementAt(0).Dependency.Dependency, Is.Not.SameAs(result.Dependencies.ElementAt(1).Dependency.Dependency));
+        }
+
+        Assert.That(result.Dependencies.ElementAt(0).Singleton, Is.SameAs(result.Dependencies.ElementAt(1).Singleton));
+    }
+
+    class Context : ScenarioContext
+    {
+        public ConcurrentBag<IDependency> Dependencies { get; } = [];
+    }
+
+    // Custom component that mimicks registering global shared keyed services
+    class CustomComponent : ComponentRunner, IComponentBehavior
```

Comment:
Nitpick: I struggled with the name, perhaps `ComponentThatRegistersGloballySharedServices`?